src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,831 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+   }
+
+   seq_dec.end_cons();
+}
+
+void IPAddressBlocks::sort_and_merge() {
+   // Sort IPAddressFamilies by afi/safi values
+   //
+   // see: https://www.rfc-editor.org/rfc/rfc3779.html#section-2.2.3.3
+   //
+   // families with no safis are ordered before families with safis,
+   // v4 families are ordered before v6 families (i.e. they are sorted by afis)
+   //
+   // families with the same afi/safi combination are then merged
+
+   // std::map is ordered, so a uint32_t of [afi,safi] will be in the right order
+   std::map<uint32_t, std::vector<IPAddressFamily>> afam_map;
+   for(const IPAddressFamily& block : m_ip_addr_blocks) {
+      uint32_t afam = block.afi();
+      if(block.safi().has_value()) {
+         afam = static_cast<uint32_t>(afam << 8) | block.safi().value();
+      }
+      std::vector<IPAddressFamily>& fams = afam_map[afam];
+      fams.push_back(block);
+   }
+
+   std::vector<IPAddressFamily> merged_blocks;
+   for(auto it = afam_map.begin(); it != afam_map.end(); it++) {
+      // fams consists of families with the same afi/safi combination
+      std::vector<IPAddressFamily>& fams = it->second;
+      // since at least 1 block has to belong to a afi/safi combination for it to appear in the map,
+      // fams cannot be empty
+      BOTAN_ASSERT_NOMSG(!fams.empty());
+
+      // fams[0] has to have the same choice type as the fams in the same bucket
+      if(std::holds_alternative<IPAddressChoice<Version::IPv4>>(fams[0].addr_choice())) {
+         merged_blocks.push_back(merge<Version::IPv4>(fams));
+      } else {
+         merged_blocks.push_back(merge<Version::IPv6>(fams));
+      }
+   }
+   m_ip_addr_blocks = merged_blocks;
+}
+
+template <IPAddressBlocks::Version V>
+IPAddressBlocks::IPAddressFamily IPAddressBlocks::merge(std::vector<IPAddressFamily>& blocks) {
+   // Merge IPAddressFamilies that have the same afi/safi combination
+   //
+   // see: https://www.rfc-editor.org/rfc/rfc3779.html#section-2.2.3.3
+
+   BOTAN_ASSERT(!blocks.empty(), "Cannot merge an empty set of IP address blocks into a single family");
+
+   // nothing to merge
+   if(blocks.size() == 1) {
+      return blocks[0];
+   }
+
+   bool all_inherit = true;
+   bool none_inherit = true;
+   for(IPAddressFamily& block : blocks) {
+      IPAddressChoice<V> choice = std::get<IPAddressChoice<V>>(block.addr_choice());
+      all_inherit = !choice.ranges().has_value() && all_inherit;  // all the blocks have the 'inherit' value
+      none_inherit = choice.ranges().has_value() && none_inherit;
+   }
+
+   // they are all 'inherit', short-circuit using default constructor for nullopt
+   if(all_inherit) {
+      return IPAddressFamily(IPAddressChoice<V>(), blocks[0].safi());
+   }
+
+   // some are inherit, and some have values - no sensible way to merge them
+   if(!all_inherit && !none_inherit) {
+      throw Exception("Only one of \"inherit\" / \"don't inherit\" is allowed per family");
```

Comment:
This should be a `Decoding_Error`

Reply:
And probably with some prefix to give further hints to someone trying to debug this from a log eg "Invalid IpAddressBlocks: "